File: frontend_old/js/downloadManager.js
/* ==========================================================
   Easy AI Studio
   File: frontend/js/downloadManager.js
   Version: 1.0.0
   Sprint: 5 - Download Manager
   ========================================================== */

"use strict";

/* ==========================================================
   Download Manager
   ========================================================== */

const DownloadManager = {

    downloads: [],

    active: null,

    status: "idle",

    /* ======================================================
       Initialize
       ====================================================== */

    init() {

        console.log(

            "Download Manager Ready"

        );

    },

    /* ======================================================
       Add Download
       ====================================================== */

    add(options = {}) {

        const download = {

            id: Utils.uuid(),

            name: options.name || "Download",

            url: options.url || "",

            size: options.size || 0,

            downloaded: 0,

            speed: 0,

            progress: 0,

            status: "queued",

            created: Utils.now()

        };

        this.downloads.push(download);

        Notification.info(

            "Download added to queue."

        );

        this.start();

        return download;

    },

    /* ======================================================
       Start Queue
       ====================================================== */

    async start() {

        if (

            this.status === "running"

        ) {

            return;

        }

        const next = this.downloads.find(

            item =>

            item.status === "queued"

        );

        if (!next) {

            return;

        }

        this.status = "running";

        this.active = next;

        next.status = "downloading";

        for (

            let progress = 0;

            progress <= 100;

            progress += 5

        ) {

            next.progress = progress;

            next.downloaded =

                (next.size / 100) * progress;

            next.speed =

                Utils.random(

                    5,

                    60

                );

            await Utils.delay(200);

        }

        next.status = "completed";

        next.progress = 100;

        next.speed = 0;

        Notification.success(

            `${next.name} completed.`

        );

        this.active = null;

        this.status = "idle";

        this.start();

    },

    /* ======================================================
       Cancel
       ====================================================== */

    cancel(id) {

        const item = this.get(id);

        if (!item) {

            return;

        }

        item.status = "cancelled";

        item.speed = 0;

        Notification.warning(

            "Download cancelled."

        );

    },

    /* ======================================================
       Find
       ====================================================== */

    get(id) {

        return this.downloads.find(

            item =>

            item.id === id

        );

    },

    /* ======================================================
       Active
       ====================================================== */

    current() {

        return this.active;

    },

    /* ======================================================
       All Downloads
       ====================================================== */

    all() {

        return this.downloads;

    },

    /* ======================================================
       Clear Completed
       ====================================================== */

    clearCompleted() {

        this.downloads =

            this.downloads.filter(

                item =>

                item.status !== "completed"

            );

    }

};

/* ==========================================================
   Global
   ========================================================== */

window.DownloadManager = DownloadManager;

File: backend/services/download_manager.py
"""
==========================================================
Easy AI Studio
File    : backend/services/download_manager.py
Version : 1.0.0
Sprint  : 10
==========================================================
"""

from pathlib import Path
from typing import List


class DownloadManager:

    def __init__(self):

        self.queue = []

    def add(

        self,

        url: str,

        output: str

    ) -> dict:

        item = {

            "url": url,

            "output": output,

            "status": "queued"

        }

        self.queue.append(item)

        return item

    def all(self) -> List[dict]:

        return self.queue

    def get(

        self,

        index: int

    ):

        if 0 <= index < len(self.queue):

            return self.queue[index]

        return None

    def remove(

        self,

        index: int

    ) -> bool:

        if 0 <= index < len(self.queue):

            self.queue.pop(index)

            return True

        return False

    def clear(self):

        self.queue.clear()

    def count(self) -> int:

        return len(self.queue)

    def mark_running(

        self,

        index: int

    ):

        item = self.get(index)

        if item:

            item["status"] = "running"

    def mark_finished(

        self,

        index: int

    ):

        item = self.get(index)

        if item:

            item["status"] = "finished"

    def mark_failed(

        self,

        index: int

    ):

        item = self.get(index)

        if item:

            item["status"] = "failed"

    def output_exists(

        self,

        path: str

    ) -> bool:

        return Path(path).exists()


download_manager = DownloadManager()